### PR TITLE
Bug/2998 null not working in q and mq filter

### PR DESCRIPTION
**Pull request by kzangeli**
Fix for issue #2998

A temporal change, related to issue https://github.com/telefonicaid/fiware-orion/issues/2587 had to be made for the fix to work.
This temporal change is a bugfix, but not the *definitive bugfix*.

### CODE REVIEW

File: test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+--SHELL--
+
+#
+# 01. Create subscription for entity E*, attribute 't' with a filter mq=t.m==null
+# 02. Create entity E1 with attr 't' having a metadata 'm' == null
+# 03. Make sure ONE notification has been sent
+# 04. Modify E1::t.m to 4
+# 05. Make sure no notification has been sent
+# 06. Modify E1::t.m to null
+# 07. Make sure ONE notification has been sent
```
Comment (fgalan) [inline test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test]: The following for 'null as string' could be added to make the test stronger:

```
# 07. Modify E1::t.m to 'null'
# 08. Make sure no notification has been sent
```

File: test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_q_for_subscription.test [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+#
+# 01. Create subscription for entity E*, attribute 't' with a filter q=t==null
+# 02. Create entity E1 with attr 't' == null
+# 03. Make sure ONE notification has been sent
+# 04. Modify E1::t to 4
+# 05. Make sure no notification has been sent
+# 06. Modify E1::t to null
+# 07. Make sure ONE notification has been sent
+#
```
Comment (fgalan) [inline test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_q_for_subscription.test]: The following for 'null as string' could be added to make the test stronger:

```
# 07. Modify E1::t to 'null'
# 08. Make sure no notification has been sent
```

File: test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test [hunk 3]
```diff
@@ -0,0 +1,258 @@
+# Copyright 2017 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test]: We are aproaching to 1000 .test (wow! :O). Please, modify testHarness.sh in this PR to add a new digit to the counter.

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 4]
```diff
@@ -2274,9 +2274,19 @@ static void updateAttrInNotifyCer
       //
       // FIXME P6: https://github.com/telefonicaid/fiware-orion/issues/2587
       // If an attribute has no value, then its value is not updated (neither is previousValue).
-      // However this may be problematic ... see the issue
+      // However this may be problematic ... see the issue.
       //
+      // New data on this: the functest "null_not_working_in_q_for_subscription.test" fails with this 'if-clause' (not outdeffed),
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: case/XXXX/null_not_working_in_q_for_subscription.test
case/XXXX/only q_and_mq_as_uri_param_for_metadata.test

will help to locate these .test files in the future

(I mean just casesXXX/, I think case/XXXX_long_name_sometimes_very_long would make thinks worse)

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 5]
```diff
@@ -2274,9 +2274,19 @@ static void updateAttrInNotifyCer
       //
       // FIXME P6: https://github.com/telefonicaid/fiware-orion/issues/2587
       // If an attribute has no value, then its value is not updated (neither is previousValue).
-      // However this may be problematic ... see the issue
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: CNR line regarding this PR missing. Suggestion:

```
Fix: null not working in q/mq filter in subscriptions (#2998)
```

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 4 above)
Comment (kzangeli) [reply src/lib/mongoBackend/MongoCommonUpdate.cpp]: I assumed people would be bright enough to issue a
`find . -name null_not_working_in_q_for_subscription.test`.
You think not? :-)

Sure, I can add indications on how to find the file

File: test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test (hunk 1 above)
Comment (kzangeli) [reply test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test]: Fixed in ba413e82e

File: test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_q_for_subscription.test (hunk 2 above)
Comment (kzangeli) [reply test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_q_for_subscription.test]: Fixed in ba413e8

File: test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test (hunk 3 above)
Comment (kzangeli) [reply test/functionalTest/cases/2998_null_not_working_in_q_and_mq_filter/null_not_working_in_mq_for_subscription.test]: Fixed in ba413e8

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 4 above)
Comment (kzangeli) [reply src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in ba413e8

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 5 above)
Comment (kzangeli) [reply src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in ba413e8

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 6]
```diff
@@ -2274,9 +2274,20 @@ static void updateAttrInNotifyCer
...
+      // However this may be problematic ... see the issue.
       //
+      // New data on this: the functest
+      // "test/functionalTest/cases/2998*/null_not_working_in_q_for_subscription.test" fails with this 'if-clause' (not outdeffed),
+      // and removing the 'if' the functest 'only q_and_mq_as_uri_param_for_metadata.test' fails, but it seems like the test is
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: cases/xxxx*/q_and_mq_as_uri_param_for_metadata.test also ;)

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 4 above)
Comment (fgalan) [reply src/lib/mongoBackend/MongoCommonUpdate.cpp]: Of course user can use grep or any other similar tool in its operations system but... why don't provide extra help in the comment itself? ;)

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 6 above)
Comment (kzangeli) [reply src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fixed in 9b203a5a6

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 7]
```diff
@@ -2274,9 +2274,20 @@ static void updateAttrInNotifyCer
       //
       // FIXME P6: https://github.com/telefonicaid/fiware-orion/issues/2587
       // If an attribute has no value, then its value is not updated (neither is previousValue).
-      // However this may be problematic ... see the issue
+      // However this may be problematic ... see the issue.
       //
+      // New data on this: the functest
+      // "test/functionalTest/cases/2998*/null_not_working_in_q_for_subscription.test" fails with this 'if-clause' (not outdeffed),
+      // and removing the 'if' the functest 'test/functionalTest/cases/2998*/only q_and_mq_as_uri_param_for_metadata.test' fails,
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: 'test/functionalTest/cases/2998*/only q_and_mq_as_uri_param_for_metadata.test'

to

'test/functionalTest/cases/1156*/q_and_mq_as_uri_param_for_metadata.test'

(I guess. Check it, please)

Comment (kzangeli) [reply src/lib/mongoBackend/MongoCommonUpdate.cpp]: Lapsus ...
Fixed in bcf30f111 (I certainly hope! :-))

**Comment by fgalan** [comment]
LGTM

**Pull request merged**